unit-tests/py/rspy/devices.py

```diff
@@ -93,6 +93,75 @@ def enabled( self ):
         return self._removed is False
 
 
+def disable_all( timeout = 5):
+    """
+    waits for all ports to be disabled
+    """
+    for retry in range( timeout ):
+        if len( enabled() ) == 0:
+            break
+        time.sleep( 1 )
+    if len( enabled() ) != 0:
+        log.w( 'Could not disable ports, can\'t infer ports for recovery devices' )
+        return
+
+
+def map_devices_with_unknown_ports():
+    """
+    Fill in unknown ports in devices by enabling one port at a time, finding out which device
+    is there.
+    """
+    global _device_by_sn
+    try:
+        devices_with_unknown_ports = [device for device in _device_by_sn.values() if device.port is None]
+        if devices_with_unknown_ports:
+            log.d( 'trying do discover unknown ports' )
+            ports = acroname.ports()
+            unknown_ports = [port for port in ports if port not in known_ports]
+            log.debug_indent()
+            log.d( "active ports:", ports )
+            log.d( "- known ports:", known_ports )
+            log.d( "= unknown ports:", unknown_ports )
+            for known_port in known_ports:
+                if known_port not in ports:
+                    log.e( "A device was found on port", known_port, "but the port is not reported as used by Acroname!" )
+            if len( unknown_ports ) == 1:
+                device = devices_with_unknown_ports[0]
+                log.d( 'port', unknown_ports[0], 'has device', device.serial_number )
+                device._port = unknown_ports[0]
+            else:
+                # disabling all ports
+                acroname.disable_ports( ports )
+                disable_all()
+                # enabling one port at a time to try and find what device is connected to it
+                n_identified_ports = 0
+                for port in unknown_ports:
+                    acroname.enable_ports( [port], disable_other_ports=True )
+                    sn = None
+                    for retry in range( 5 ):
+                        if len( enabled() ) == 1:
+                            sn = list( enabled() )[0]
+                            break
+                        time.sleep( 1 )
+                    if not sn:
+                        log.d( 'could not recognise device in port', port )
+                    else:
+                        device = _device_by_sn.get( sn )
+                        if device:
+                            log.d( 'port', port, 'has device', sn )
+                            device._port = port
+                            n_identified_ports += 1
+                            if len( devices_with_unknown_ports ) == n_identified_ports: # identified ports of all devices
+                                break
+                        else:
+                            log.w( "Device with serial number", sn, "was found in port", port,
```

Comment:
If the SN is not in `_device_by_sn`, then `get(sn)` will throw

Reply:
I checked that. It doesn't, it returns None

Reply:
Hmm you're right... for some reason I thought otherwise